test/parallel/test-vm-module-dynamic-import.js

```diff
@@ -19,8 +17,51 @@ const finished = common.mustCall();
     await result;
   } catch (err) {
     threw = true;
-    assert.strictEqual(err.message, 'import() called outside of main context');
+    assert.strictEqual(err.code, 'ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING');
   }
   assert(threw);
-  finished();
-}());
+}
+
+async function test() {
+  const foo = new SourceTextModule('export const a = 1;');
+  await foo.link(common.mustNotCall());
+  foo.instantiate();
+  await foo.evaluate();
+
+  {
+    const s = new Script('import("foo")', {
+      importModuleDynamically: common.mustCall((specifier, wrap) => {
+        assert.strictEqual(specifier, 'foo');
+        assert.strictEqual(wrap, s);
+        return foo.namespace;
```

Comment:
Can we have some cases of invalid returns being tested here? What is the error?

Reply:
there is no limit on what you can return.

Reply:
This seems very dangerous (and I'm sure you'll agree!).

As an alternative could we not do something like:
* Only support a vm.Module instance from importModuleDynamically
* To delegate to `loader.import` have a special "loader" return value or indicator for this function.

Reply:
I don't really see any danger, could you fill me in on how you're approaching this?

Reply:
Well say I'm using this API for the first time - I will likely just return an object here and see that it works and make that the pattern I use in my code for dynamic imports. We shouldn't be encouraging a model that dynamic import is just "anything", but rather that it is another linking process in the graph. There's a big difference between exposing a hook that can return anything and a hook that goes through the same module logic again. So I'm just saying we should probably have it take `vm.Module` instances as that is the module logic.

For example, error handling should be the same between environments - in this approach an error throw in dynamic import resolution isn't cached. That is a spec violation.

Reply:
it's definitely not a linking process, that's why I removed the context check in the source.

I think all we end up doing with limitations is cutting out options. For instance if someone exports a thenable namespace, it might be resolved at any point in someone's loader chain before being returned into the dynamic import handler.

Reply:
As an alternative we could make the user return an object with a property pointing to the namespace object. This way we could typecheck the return value and avoid the whole thenable business until it is necessary.

Reply:
If we do check the type I would like it to only be an "is namespace" check, because returning a SourceTextModule instance doesn't guarantee that the SourceTextModule instance is in the correct state to be giving out namespaces, so i think its better to explicitly force the user to get that far. It's also worth mentioning that checking if a value is a module namespace is rather expensive because it's a native call.

Reply:
Yes a namespace check is fine.

Reply:
To try and clarify some of the cases here:

* Having `import('x')` twice in the same source can return different modules each time it is called. This isn't explicitly a spec invariant, so could be seen as acceptable.
* It is possible for `import('x')` to provide an error that does not correspond to an instantiation or evaluation error on a module record.
* There are no guarantees that `import('x')` when resolving to the same module twice, will return the same cached error

With the namespace type check:
* It is possible for `import('x')` to not be able to resolve a valid module that is a thenable
Without the namespace type check:
* It is possible for `import('x')` to return any JavaScript binding without invoking any module loading algorithms at all.

If we mandated a return value type of "Module | "defaultLoader" (<string>) (not final API but the API suggestion, then the cases become:
* Once the promise has resolved to the "module" | default loader indicator string, we then get correct module errors and caching. Errors before this resolution that are not instantiation errors can still be spec-invariant, but we close this gap somewhat.
* It is possible to resolve thenable modules
* It is not possible to resolve an arbitrary JS binding without resolving through a Module